lib/iris/__init__.py

```diff
@@ -91,13 +91,10 @@ def callback(cube, field, filename):
 import os.path
 import threading
 
-import iris.config
-import iris.cube
```

Comment:
I'm totally happy for this to be in a minor release. 

Assuming `iris.cube.Cube` is in the root namespace is a pretty weak assumption, and it's certainly not part of the API or any contact with the user. Nor is it there by design.

Also, users can't complain about slow imports and also want the convenience of most things being freely imported and available in the namespace.

Like you said, just import if you want it.

Reply:
Also, thinking about this a wee bit more... If this PR had removed an entry from the `iris.__all__`, which I guess is as close to a module namespace contract that you can have in Python. Then one could argue that should be deemed as a breaking change. However, that's not the case here. 

I get your point @rcomer and I don't want to be unreasonable, but I'm convincing myself further that I'm happy with this change.

Perhaps other @SciTools/iris-devs have an opinion on this...

Reply:
>Assuming `iris.cube.Cube` is in the root namespace is a pretty weak assumption, and it's certainly not part of the API or any contract with the user. Nor is it there by design.

I hate to bring this up _again_, but [xkcd#1172](https://xkcd.com/1172/) seems apt here. While we will undoubtedly break some people's workflows, it will only be those who had constructed workflows based on unintended un-contracted behaviour.

Reply:
Thanks both.  As long as the question has been considered and we're happy then that's good enough for me.  I'm just keen to avoid situations where devs make changes without realising they will break user code - that has happened before (and is what prompted me to start checking GH on a frequent basis...)

Now I think about it, I remember being pretty confused when I started using python/Iris that I had to import some modules but not others.  Guess I just accepted it and stopped thinking about it at some point.

Reply:
On one hand, I've literally never written `import iris.cube`, but generally do assume I can at least create a new `iris.cube.CubeList`.

On the other... in practice it'll probably break a lot less than you expect.

That is, while this won't work:
```python
import iris
cube = iris.cube.Cube([])
```
this should:
```python
import iris
cubes = iris.load("some file.nc")
cube = iris.cube.Cube([])
```

Because `iris.load` clearly imports `iris.cube` somewhere along the line, which will modify *the* `iris` module object (there is only ever one) in-place with the new `cube` submodule.

But back on the first hand, if linters statically inspect code instead of importing it (I actually have no idea where they normally stand on this) then they'll still shout at you until you explicitly import `iris.cube`.

Reply:
Thanks @bsherratt, yes you are right.  Similarly if `iris.plot` has been loaded it will still work.  There's probably not much user code that doesn't do any loading or plotting.  So I'm probably worrying about nothing.

Reply:
Indeed it does...
```bash
itwl@vld173:> python
Python 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import iris
>>> cube = iris.cube.Cube([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'iris' has no attribute 'cube'
>>> cubes = iris.load(iris.sample_data_path("air_temp.pp"))
>>> cube = iris.cube.Cube([])
>>> 
```

Reply:
> Because iris.load clearly imports iris.cube somewhere along the line ...
> ... do assume I can at least create a new iris.cube.CubeList.

Those are true, but they also co-incidental : it could change.
And doing so may break people's code (it has happened before, possibly with `iris.analysis.cartography` IIRC ?)

If the user writes "iris.cube.<anything>", then they "ought" to import "iris.cube".

Personally, I find these days that PyCharm will remind me of most of those.

Reply:
> If the user writes iris.cube.<anything>, then they "ought" to import "iris.cube"

Literally noone has ever told me that.  This thread has been an education!

Reply:
> If the user writes `iris.cube.<anything>`, then they "ought" to `import "iris.cube"`.

Well, yes and no... as you say, no rules, even within the standard library. For example `urllib` - you get `urllib.parse` for "free", but the other submodules need to be imported explicitly. Or `os.path`, which *changed* behaviour: used to (~3.3?) need importing explicitly, but now `import os` is sufficient. It seems pretty deliberate that packages are free to choose how many submodules get automatically exposed.

In small personal code I definitely expose all of the submodules for example, but of course the larger a package gets, the less reasonable this becomes. It's totally down to the package to decide. (Personally, I'm marginally on the side that `cube` is "core" enough that it might as well be imported upfront.)

Reply:
This is why I always start with the `from iris.cube import Cube` format  as my default, just to avoid this uncertainty.

Reply:
> marginally on the side that cube is "core" enough that it might as well be imported upfront

#891 proposed actually moving `Cube` to the top level.